polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -404,16 +441,45 @@ pub mod pallet {
 		///
 		/// Can be called by Root, the parachain, or the parachain manager if the parachain is
 		/// unlocked.
+		///
+		/// In case the call is made by the parachain manager or the parachain itself the upgrade
+		/// fee will be charged.
+		///
+		/// The caller will receive a refund or be required to reserve an additional deposit,
+		/// depending on the size of the new validation code.
 		#[pallet::call_index(7)]
 		#[pallet::weight(<T as Config>::WeightInfo::schedule_code_upgrade(new_code.0.len() as u32))]
 		pub fn schedule_code_upgrade(
 			origin: OriginFor<T>,
 			para: ParaId,
 			new_code: ValidationCode,
 		) -> DispatchResult {
-			Self::ensure_root_para_or_owner(origin, para)?;
-			runtime_parachains::schedule_code_upgrade::<T>(para, new_code, SetGoAhead::No)?;
-			Ok(())
+			Self::ensure_root_para_or_owner(origin.clone(), para)?;
+
+			let fee_payer = if ensure_root(origin.clone()).is_ok() {
+				// Root doesn't pay. This, also means that system parachains do not pay for upgrade
+				// fees.
+				None
+			} else if LegacyParas::<T>::get(para).is_some() {
+				// Each legacy para is permitted to perform one upgrade for free.
+				//
+				// This is introduced to  avoid causing a breaking change to the system once para
+				// upgrade fees are required.
+				LegacyParas::<T>::remove(para);
+				None
+			} else if let Ok(caller) = ensure_signed(origin) {
+				let para_info = Paras::<T>::get(para).ok_or(Error::<T>::NotRegistered)?;
+				ensure!(!para_info.is_locked(), Error::<T>::ParaLocked);
+				ensure!(para_info.manager == caller, Error::<T>::NotOwner);
+
+				Some(caller)
+			} else {
+				let location: MultiLocation = (Parent, Parachain(para.into())).into();
+				let sovereign_account = T::SovereignAccountOf::convert_location(&location).unwrap();
```

Comment:
I think we should make it so that in order to do any upgrades you have to manually configure an account paying for fees and providing the deposit(changes). This way, if an parachain has not set this up, updates will simply fail.

Otherwise, teams not reading documentation might be surprised where all their funds went ...

For now, I think the best option is to force users/projects to manually configure the paying/depositing account in order for upgrades to succeed, e.g. we should offer an extrinsic: "set_upgrade_fee_and_deposit_paying_account" and upgrades will only succeed if this extrinsic has been called before and the set account is sufficiently funded (obviously). 

We could make this a one-shot configuration as part of the PVF registration, but I think it has to be explicit. On the other hand (not part of this PR), we should also have crystal clear documentation (or even better tooling), so users can determine what their upgrade will cost.